##### Nama: Nurul Fatimah
##### Submission: Recommendation System

###### Import Data

In [77]:
# Mengimport library pandas
import pandas as pd

# Mengimport modul drive untuk akses ke google drive
from google.colab import drive

# Mengakses Google Drive
drive.mount('/content/drive')

# Membuat file_path yang akan diakses
books = '/content/drive/My Drive/dataset/book/Books.csv'
ratings = '/content/drive/My Drive/dataset/book/Ratings.csv'
users = '/content/drive/My Drive/dataset/book/Users.csv'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


###### Membaca dan Menampilkan Data

In [78]:
# Membaca file path buku
df_books = pd.read_csv(books)

# Menampilkan data buku
df_books

<ipython-input-78-eb93004440be>:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  df_books = pd.read_csv(books)


ISBN                                         Book-Title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271355  0440400988                         There's a Bat in Bunk Five   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                 Book-Author Year-Of-Publication  \
0         Mark P. O. Morford                2002   
1       Richard Bruce Wright                2001   
2               Carlo D'Este                1991   
3           Gina Bari Kolata                1999   
4            E. J. W. Barber                1999   
...                      ...                 ...   
271355        Paula Danziger                1988   
271356            Teri Sloat                1991   
271357      Christine Wicker                2004   
271358                 Plato                1996   
271359   Christopher  Biffle                2000   

                                               Publisher  \
0                                Oxford University Press   
1                                  HarperFlamingo Canada   
2                                        HarperPerennial   
3                                   Farrar Straus Giroux   
4                             W. W. Norton &amp; Company   
...                                                  ...   
271355                   Random House Childrens Pub (Mm)   
271356                                      Dutton Books   
271357                                HarperSanFrancisco   
271358                           Oxford University Press   
271359  McGraw-Hill Humanities/Social Sciences/Languages   

                                              Image-URL-S  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-M  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-L  
0       http://images.amazon.com/images/P/0195153448.0...  
1       http://images.amazon.com/images/P/0002005018.0...  
2       http://images.amazon.com/images/P/0060973129.0...  
3       http://images.amazon.com/images/P/0374157065.0...  
4       http://images.ama

Penjelasan dari setiap column di book :
- ISBN: Nomor Standar Buku Internasional yang unik untuk setiap buku. Ini adalah kunci utama untuk mengidentifikasi buku.
- Book-Title: Judul buku.
- Book-Author: Nama penulis buku.
- Year-Of-Publication: Tahun publikasi buku.
- Publisher: Penerbit buku.
- Image-URL-S: URL untuk gambar sampul buku dengan ukuran kecil.
- Image-URL-M: URL untuk gambar sampul buku dengan ukuran sedang.
- Image-URL-L: URL untuk gambar sampul buku dengan ukuran besar.

In [79]:
#Membaca file path rating
df_ratings = pd.read_csv(ratings)

#Menampilkan data rating
df_ratings

User-ID         ISBN  Book-Rating
0         276725   034545104X            0
1         276726   0155061224            5
2         276727   0446520802            0
3         276729   052165615X            3
4         276729   0521795028            6
...          ...          ...          ...
1149775   276704   1563526298            9
1149776   276706   0679447156            0
1149777   276709   0515107662           10
1149778   276721   0590442449           10
1149779   276723  05162443314            8

[1149780 rows x 3 columns]

Penjelasan dari setiap column di ratings:
- User-ID: ID unik untuk setiap pengguna. Ini berfungsi sebagai kunci utama untuk mengidentifikasi pengguna.
- ISBN: Merujuk ke ISBN dalam tabel buku, menghubungkan penilaian dengan buku yang spesifik.
- Book-Rating: Nilai atau rating yang diberikan oleh pengguna untuk buku tersebut, biasanya dalam skala 1 hingga 10.

In [80]:
# Membaca file path pengguna
df_users = pd.read_csv(users)

# Menampilkan data pengguna
df_users

User-ID                            Location   Age
0             1                  nyc, new york, usa   NaN
1             2           stockton, california, usa  18.0
2             3     moscow, yukon territory, russia   NaN
3             4           porto, v.n.gaia, portugal  17.0
4             5  farnborough, hants, united kingdom   NaN
...         ...                                 ...   ...
278853   278854               portland, oregon, usa   NaN
278854   278855  tacoma, washington, united kingdom  50.0
278855   278856           brampton, ontario, canada   NaN
278856   278857           knoxville, tennessee, usa   NaN
278857   278858                dublin, n/a, ireland   NaN

[278858 rows x 3 columns]

Penjelasan column pada users:
- User-ID: ID unik untuk setiap pengguna, yang juga digunakan dalam tabel ratings untuk mengidentifikasi siapa yang memberi rating.
- Location: Lokasi geografis pengguna, biasanya berupa kota, negara bagian, dan negara.
- Age: Usia pengguna.

###### Exploratory Data Analysis

* Melihat Informasi Data

In [81]:
# Informasi data
print('='*19)
print('Informasi Data Buku')
print('='*19)
df_books.info()
print('\n')
print('='*21)
print('Informasi Data Rating')
print('='*21)
df_ratings.info()
print('\n')
print('='*23)
print('Informasi Data Pengguna')
print('='*23)
df_users.info()

Informasi Data Buku
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271358 non-null  object
 3   Year-Of-Publication  271360 non-null  object
 4   Publisher            271358 non-null  object
 5   Image-URL-S          271360 non-null  object
 6   Image-URL-M          271360 non-null  object
 7   Image-URL-L          271357 non-null  object
dtypes: object(8)
memory usage: 16.6+ MB


Informasi Data Rating
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1149780 entries, 0 to 1149779
Data columns (total 3 columns):
 #   Column       Non-Null Count    Dtype 
---  ------       --------------    ----- 
 0   User-ID      1149780 non-null  int64 
 1   ISBN         1149780 non-null  object
 2   Book-Rating  114

Dataset ini terdiri dari tiga tabel dengan jumlah data dan tipe data sebagai berikut:

1. **Data Buku**: Terdiri dari 271,360 entri, semua kolom bertipe `object` (string).
   
2. **Data Rating**: Memiliki 1,149,780 entri, dengan dua kolom bertipe `int64` (`User-ID` dan `Book-Rating`) dan satu kolom bertipe `object` (`ISBN`).

3. **Data Pengguna**: Terdiri dari 278,858 entri, dengan kolom `User-ID` bertipe `int64`, `Location` bertipe `object`, dan `Age` bertipe `float64`.

* Melihat Statistik Data

In [82]:
# Melihat statistik data buku
print("Statistik deskriptif dari data buku: ")
df_books.describe(include='all')

Statistik deskriptif dari data buku: 


ISBN      Book-Title      Book-Author  Year-Of-Publication  \
count       271360          271360           271358               271360   
unique      271360          242135           102022                  202   
top     0195153448  Selected Poems  Agatha Christie                 2002   
freq             1              27              632                13903   

        Publisher                                        Image-URL-S  \
count      271358                                             271360   
unique      16807                                             271044   
top     Harlequin  http://images.amazon.com/images/P/185326119X.0...   
freq         7535                                                  2   

                                              Image-URL-M  \
count                                              271360   
unique                                             271044   
top     http://images.amazon.com/images/P/185326119X.0...   
freq                                                    2   

                                              Image-URL-L  
count                                              271357  
unique                                             271041  
top     http://images.amazon.com/images/P/225307649X.0...  
freq                                                    2

Statistik deskriptif dari data buku menunjukkan bahwa dataset ini terdiri dari 271,360 entri unik untuk ISBN, judul buku, dan URL gambar sampul (baik ukuran kecil, sedang, maupun besar). Terdapat 242,135 judul buku unik dan 102,022 penulis unik. Tahun publikasi paling sering adalah tahun 2002 dengan frekuensi 13,903 buku, sementara penerbit yang paling banyak terdaftar adalah Harlequin, yang muncul sebanyak 7,535 kali. Beberapa ISBN dan gambar sampul buku memiliki duplikasi, dengan frekuensi tertinggi adalah 2. Buku dengan judul "Selected Poems" oleh Agatha Christie adalah yang paling sering muncul, sebanyak 27 kali dalam dataset.

In [83]:
# Melihat statistik data rating
print("\nStatistik deskriptif dari data rating: ")
df_ratings.describe(include='all')


Statistik deskriptif dari data rating: 


User-ID        ISBN   Book-Rating
count   1.149780e+06     1149780  1.149780e+06
unique           NaN      340556           NaN
top              NaN  0971880107           NaN
freq             NaN        2502           NaN
mean    1.403864e+05         NaN  2.866950e+00
std     8.056228e+04         NaN  3.854184e+00
min     2.000000e+00         NaN  0.000000e+00
25%     7.034500e+04         NaN  0.000000e+00
50%     1.410100e+05         NaN  0.000000e+00
75%     2.110280e+05         NaN  7.000000e+00
max     2.788540e+05         NaN  1.000000e+01

Statistik deskriptif dari data rating menunjukkan bahwa dataset ini terdiri dari 1,149,780 entri rating. Kolom **User-ID** memiliki rata-rata nilai sekitar 140,386, dengan rentang dari 2 hingga 278,854. Kolom **ISBN** mencatat 340,556 ISBN unik, dengan ISBN "0971880107" muncul paling sering, yaitu sebanyak 2,502 kali. **Book-Rating** yang diberikan oleh pengguna berkisar dari 0 hingga 10, dengan nilai rata-rata sekitar 2.87 dan standar deviasi 3.85. Sebagian besar buku memiliki rating 0, dengan nilai kuartil pertama, kedua (median), dan ketiga masing-masing 0, 0, dan 7.

In [84]:
# Melihat statistik data pengguna
print("\nStatistik deskriptif dari data pengguna: ")
df_users.describe(include='all')


Statistik deskriptif dari data pengguna: 


User-ID                         Location            Age
count   278858.00000                           278858  168096.000000
unique           NaN                            57339            NaN
top              NaN  london, england, united kingdom            NaN
freq             NaN                             2506            NaN
mean    139429.50000                              NaN      34.751434
std      80499.51502                              NaN      14.428097
min          1.00000                              NaN       0.000000
25%      69715.25000                              NaN      24.000000
50%     139429.50000                              NaN      32.000000
75%     209143.75000                              NaN      44.000000
max     278858.00000                              NaN     244.000000

Statistik deskriptif dari data pengguna menunjukkan bahwa dataset ini terdiri dari 278,858 entri pengguna unik. Kolom **Location** mencatat 57,339 lokasi unik, dengan "london, england, united kingdom" sebagai lokasi yang paling sering muncul, yaitu sebanyak 2,506 kali. Kolom **Age** memiliki data untuk 168,096 pengguna, dengan usia rata-rata sekitar 34.75 tahun dan standar deviasi 14.43 tahun. Usia pengguna berkisar dari 0 hingga 244 tahun, dengan nilai kuartil pertama, kedua (median), dan ketiga masing-masing 24, 32, dan 44 tahun. Kolom **User-ID** bervariasi dari 1 hingga 278,858, dengan rata-rata sekitar 139,429.5.

* Cek Missing Value

In [85]:
# Cek missing value dalam data buku
print('Cek missing value dalam data buku: ')
df_books.isnull().sum()

Cek missing value dalam data buku: 


ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

Terdapat nilai kosong dalam column Book-Author sebanyak 2, Publisher sebanyak 2, dan Image-URL-L sebanyak 3 raw.

In [86]:
# Cek missing value dalam data rating
print('Cek missing value dalam data rating: ')
df_ratings.isnull().sum()

Cek missing value dalam data rating: 


User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [87]:
# Cek missing value dalam data pengguna
print('Cek missing value dalam data pengguna: ')
df_users.isnull().sum()

Cek missing value dalam data pengguna: 


User-ID          0
Location         0
Age         110762
dtype: int64

Terdapat nilai kosong pada column Age sebanyak 110762.

* Cek Duplikat

In [88]:
 # Cek data duplikat dalam data buku
 print('Cek data duplikat dalam data buku: ')
 df_books.duplicated().sum()

Cek data duplikat dalam data buku: 


0

In [89]:
 # Cek data duplikat dalam data rating
 print('Cek data duplikat dalam data rating: ')
 df_ratings.duplicated().sum()

Cek data duplikat dalam data rating: 


0

In [90]:
 # Cek data duplikat dalam data buku
 print('Cek data duplikat dalam data pengguna: ')
 df_users.duplicated().sum()

Cek data duplikat dalam data pengguna: 


0

###### Data Preprocessing

* Mengetahui Jumlah Rating

In [91]:
books_rating = df_ratings.merge(df_books, on='ISBN', how='left')
books_rating

User-ID         ISBN  Book-Rating  \
0         276725   034545104X            0   
1         276726   0155061224            5   
2         276727   0446520802            0   
3         276729   052165615X            3   
4         276729   0521795028            6   
...          ...          ...          ...   
1149775   276704   1563526298            9   
1149776   276706   0679447156            0   
1149777   276709   0515107662           10   
1149778   276721   0590442449           10   
1149779   276723  05162443314            8   

                                                Book-Title        Book-Author  \
0                                     Flesh Tones: A Novel         M. J. Rose   
1                                         Rites of Passage         Judith Rae   
2                                             The Notebook    Nicholas Sparks   
3                                           Help!: Level 1      Philip Prowse   
4        The Amsterdam Connection : Level 4 (Cambridge ...        Sue Leather   
...                                                    ...                ...   
1149775  Get Clark Smart : The Ultimate Guide for the S...       Clark Howard   
1149776  Eight Weeks to Optimum Health: A Proven Progra...        Andrew Weil   
1149777   The Sherbrooke Bride (Bride Trilogy (Paperback))  Catherine Coulter   
1149778                                  Fourth Grade Rats     Jerry Spinelli   
1149779                                                NaN                NaN   

        Year-Of-Publication                   Publisher  \
0                      2002            Ballantine Books   
1                      2001                      Heinle   
2                      1996                Warner Books   
3                      1999  Cambridge University Press   
4                      2001  Cambridge University Press   
...                     ...                         ...   
1149775                2000            Longstreet Press   
1149776                1997             Alfred A. Knopf   
1149777                1996                  Jove Books   
1149778                1996                  Scholastic   
1149779                 NaN                         NaN   

                                               Image-URL-S  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/0155061224.0...   
2        http://images.amazon.com/images/P/0446520802.0...   
3        http://images.amazon.com/images/P/052165615X.0...   
4        http://images.amazon.com/images/P/0521795028.0...   
...                                                    ...   
1149775  http://images.amazon.com/images/P/1563526298.0...   
1149776  http://images.amazon.com/images/P/0679447156.0...   
1149777  http://images.amazon.com/images/P/0515107662.0...   
1149778  http://images.amazon.com/images/P/0590442449.0...   
1149779                                                NaN   

                                               Image-URL-M  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/0155061224.0...   
2        http://images.amazon.com/images/P/0446520802.0...   
3        http://images.amazon.com/images/P/052165615X.0...   
4        http://images.amazon.com/images/P/0521795028.0...   
...                                                    ...   
1149775  http://images.amazon.com/images/P/1563526298.0...   
1149776  http://images.amazon.com/images/P/0679447156.0...   
1149777  http://images.amazon.com/images/P/0515107662.0...   
1149778  http://images.amazon.com/images/P/0590442449.0...   
1149779                                                NaN   

                                               Image-URL-L  
0        http://images.amazon.com/images/P/034545104X.0...  
1        http://images.amazon.com/images/P/0155061224.0...  
2        http://images.amazon.com/images/P/0446520802.0...  
3        http://images.amazon.com/images

Merge Data: Menggabungkan dua DataFrame, yaitu df_ratings dan df_books, berdasarkan kolom ISBN.

* Cek Missing Value books_rating

In [92]:
books_rating.isnull().sum()

User-ID                     0
ISBN                        0
Book-Rating                 0
Book-Title             118644
Book-Author            118646
Year-Of-Publication    118644
Publisher              118646
Image-URL-S            118644
Image-URL-M            118644
Image-URL-L            118648
dtype: int64

###### Data Preparation

* Membersihkan Missing Value

In [93]:
books_rating_clean = books_rating.dropna()
books_rating_clean

User-ID        ISBN  Book-Rating  \
0         276725  034545104X            0   
1         276726  0155061224            5   
2         276727  0446520802            0   
3         276729  052165615X            3   
4         276729  0521795028            6   
...          ...         ...          ...   
1149774   276704  0876044011            0   
1149775   276704  1563526298            9   
1149776   276706  0679447156            0   
1149777   276709  0515107662           10   
1149778   276721  0590442449           10   

                                                Book-Title        Book-Author  \
0                                     Flesh Tones: A Novel         M. J. Rose   
1                                         Rites of Passage         Judith Rae   
2                                             The Notebook    Nicholas Sparks   
3                                           Help!: Level 1      Philip Prowse   
4        The Amsterdam Connection : Level 4 (Cambridge ...        Sue Leather   
...                                                    ...                ...   
1149774  Edgar Cayce on the Akashic Records: The Book o...  Kevin J. Todeschi   
1149775  Get Clark Smart : The Ultimate Guide for the S...       Clark Howard   
1149776  Eight Weeks to Optimum Health: A Proven Progra...        Andrew Weil   
1149777   The Sherbrooke Bride (Bride Trilogy (Paperback))  Catherine Coulter   
1149778                                  Fourth Grade Rats     Jerry Spinelli   

        Year-Of-Publication  \
0                      2002   
1                      2001   
2                      1996   
3                      1999   
4                      2001   
...                     ...   
1149774                1998   
1149775                2000   
1149776                1997   
1149777                1996   
1149778                1996   

                                                 Publisher  \
0                                         Ballantine Books   
1                                                   Heinle   
2                                             Warner Books   
3                               Cambridge University Press   
4                               Cambridge University Press   
...                                                    ...   
1149774  A.R.E. Press (Association of Research &amp; Enlig   
1149775                                   Longstreet Press   
1149776                                    Alfred A. Knopf   
1149777                                         Jove Books   
1149778                                         Scholastic   

                                               Image-URL-S  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/0155061224.0...   
2        http://images.amazon.com/images/P/0446520802.0...   
3        http://images.amazon.com/images/P/052165615X.0...   
4        http://images.amazon.com/images/P/0521795028.0...   
...                                                    ...   
1149774  http://images.amazon.com/images/P/0876044011.0...   
1149775  http://images.amazon.com/images/P/1563526298.0...   
1149776  http://images.amazon.com/images/P/0679447156.0...   
1149777  http://images.amazon.com/images/P/0515107662.0...   
1149778  http://images.amazon.com/images/P/0590442449.0...   

                                               Image-URL-M  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/0155061224.0...   
2        http://images.amazon.com/images/P/0446520802.0...   
3        http://images.amazon.com/images/P/052165615X.0...   
4        http://images.amazon.com/images/P/0521795028.0...   
...                                                    ...   
1149774  http://images.amazon.com/images/P/0876044011.0...   
1149775  http://images.amazon.com/images/P/1563526298.0...   
1149776  http://images.amazon.com/images/P/0679447156.0...   
1149777  htt

* Mengecek Kembali Missing Value

In [94]:
books_rating_clean.isnull().sum()

User-ID                0
ISBN                   0
Book-Rating            0
Book-Title             0
Book-Author            0
Year-Of-Publication    0
Publisher              0
Image-URL-S            0
Image-URL-M            0
Image-URL-L            0
dtype: int64

* Mengecek Duplikat books_rating

In [95]:
books_rating_clean.duplicated().sum()

0

* Menghitung Jumlah Rating tiap Buku

In [96]:
jml_rate = books_rating_clean.groupby('Book-Title').count()['Book-Rating'].reset_index()
jml_rate.rename(columns={'Book-Rating':'Num_Ratings'}, inplace=True)
jml_rate

Book-Title  Num_Ratings
0        A Light in the Storm: The Civil War Diary of ...            4
1                                   Always Have Popsicles            1
2                    Apple Magic (The Collector's series)            1
3        Ask Lily (Young Women of Faith: Lily Series, ...            1
4        Beyond IBM: Leadership Marketing and Finance ...            1
...                                                   ...          ...
241060                                      Ã?Â?lpiraten.            2
241061                     Ã?Â?rger mit Produkt X. Roman.            4
241062                                Ã?Â?sterlich leben.            1
241063                              Ã?Â?stlich der Berge.            3
241064                                  Ã?Â?thique en toc            2

[241065 rows x 2 columns]

Memberikan informasi tentang seberapa banyak rating yang diberikan untuk masing-masing buku, membantu dalam analisis popularitas dan kualitas buku berdasarkan ulasan pengguna.

In [97]:
avg_rate = books_rating_clean.groupby('Book-Title')['Book-Rating'].mean().reset_index()
avg_rate.rename(columns={'Book-Rating':'Avg_Rating'}, inplace=True)
avg_rate

Book-Title  Avg_Rating
0        A Light in the Storm: The Civil War Diary of ...    2.250000
1                                   Always Have Popsicles    0.000000
2                    Apple Magic (The Collector's series)    0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...    8.000000
4        Beyond IBM: Leadership Marketing and Finance ...    0.000000
...                                                   ...         ...
241060                                      Ã?Â?lpiraten.    0.000000
241061                     Ã?Â?rger mit Produkt X. Roman.    5.250000
241062                                Ã?Â?sterlich leben.    7.000000
241063                              Ã?Â?stlich der Berge.    2.666667
241064                                  Ã?Â?thique en toc    4.000000

[241065 rows x 2 columns]

Memberikan informasi tentang rata-rata rating yang diberikan untuk masing-masing buku, yang berguna untuk analisis kualitas buku berdasarkan ulasan pengguna.

In [98]:
df_popular = jml_rate.merge(avg_rate, on='Book-Title')
df_popular

Book-Title  Num_Ratings  \
0        A Light in the Storm: The Civil War Diary of ...            4   
1                                   Always Have Popsicles            1   
2                    Apple Magic (The Collector's series)            1   
3        Ask Lily (Young Women of Faith: Lily Series, ...            1   
4        Beyond IBM: Leadership Marketing and Finance ...            1   
...                                                   ...          ...   
241060                                      Ã?Â?lpiraten.            2   
241061                     Ã?Â?rger mit Produkt X. Roman.            4   
241062                                Ã?Â?sterlich leben.            1   
241063                              Ã?Â?stlich der Berge.            3   
241064                                  Ã?Â?thique en toc            2   

        Avg_Rating  
0         2.250000  
1         0.000000  
2         0.000000  
3         8.000000  
4         0.000000  
...            ...  
241060    0.000000  
241061    5.250000  
241062    7.000000  
241063    2.666667  
241064    4.000000  

[241065 rows x 3 columns]

Kode diatas untuk menggabungkan jumlah rating dan rat-rata rating

In [99]:
df_popular = df_popular[df_popular['Num_Ratings']>250].sort_values('Avg_Rating', ascending=False).head(50)
df_popular

Book-Title  Num_Ratings  \
80431   Harry Potter and the Prisoner of Azkaban (Book 3)          428   
80419        Harry Potter and the Goblet of Fire (Book 4)          387   
80438      Harry Potter and the Sorcerer's Stone (Book 1)          278   
80423   Harry Potter and the Order of the Phoenix (Boo...          347   
80411    Harry Potter and the Chamber of Secrets (Book 2)          556   
191607  The Hobbit : The Enchanting Prelude to The Lor...          281   
187372  The Fellowship of the Ring (The Lord of the Ri...          368   
80442   Harry Potter and the Sorcerer's Stone (Harry P...          575   
211379     The Two Towers (The Lord of the Rings, Part 2)          260   
219736                              To Kill a Mockingbird          510   
183568                                  The Da Vinci Code          898   
187875                 The Five People You Meet in Heaven          430   
180552                             The Catcher in the Rye          449   
196321                          The Lovely Bones: A Novel         1295   
764                                                  1984          284   
144161                           Prodigal Summer: A Novel          253   
128667                                         Neverwhere          265   
206497                            The Secret Life of Bees          774   
168715  Stupid White Men ...and Other Sorry Excuses fo...          283   
223130  Tuesdays with Morrie: An Old Man, a Young Man,...          493   
204382                The Red Tent (Bestselling Backlist)          723   
191584               The Hitchhiker's Guide to the Galaxy          268   
129376   Nickel and Dimed: On (Not) Getting By in America          335   
93378                                       Into the Wild          252   
63864                                      Fahrenheit 451          409   
74747                           Girl with a Pearl Earring          526   
136142                                          Outlander          283   
233364  Where the Heart Is (Oprah's Book Club (Paperba...          585   
156098                     Seabiscuit: An American Legend          275   
107959                                         Life of Pi          664   
176841                                     The Bean Trees          389   
2281    A Child Called \It\": One Child's Courage to S...          265   
8433                                       ANGELA'S ASHES          279   
76340                                         Good in Bed          490   
64928   Fast Food Nation: The Dark Side of the All-Ame...          321   
12699                                       American Gods          302   
161641                                 Skipping Christmas          322   
105774  Left Behind: A Novel of the Earth's Last Days ...          318   
189546    The Golden Compass (His Dark Materials, Book 1)          336   
181675                                   The Color Purple          314   
160332                               Silence of the Lambs          256   
8751                                          About a Boy          262   
158134                  Seven Up (A Stephanie Plum Novel)          278   
175093  The Alchemist: A Fable About Following Your Dream          266   
80066   Hard Eight : A Stephanie Plum Novel (A Stephan...          269   
170097                       Suzanne's Diary for Nicholas          457   
111070                                  Lord of the Flies          259   
5664                              A Prayer for Owen Meany          413   
212065   The Vampire Lestat (Vampire Chronicles, Book II)          301   
233845       White Oleander : A Novel (Oprah's Book Club)          356   

        Avg_Rating  
80431     5.852804  
80419     5.824289  
80438     5.737410  
80423     5.501441  
80411     5.183453  
191607    5.007117  
187372    4.948370  
80442     4.895652  
211379    4.880769  
219736    4.700000  
183568    4.642539  
187875    4.551163  

Menampilkan 50 buku yang paling populer dan memiliki rating tertinggi berdasarkan jumlah rating yang diberikan, memberikan gambaran tentang buku-buku terbaik menurut ulasan pengguna.

In [100]:
books_fix = df_books.drop_duplicates('Book-Title')

In [101]:
df_books_popular = df_popular.merge(books_fix, on='Book-Title', how='left')
df_books_popular

Book-Title  Num_Ratings  \
0   Harry Potter and the Prisoner of Azkaban (Book 3)          428   
1        Harry Potter and the Goblet of Fire (Book 4)          387   
2      Harry Potter and the Sorcerer's Stone (Book 1)          278   
3   Harry Potter and the Order of the Phoenix (Boo...          347   
4    Harry Potter and the Chamber of Secrets (Book 2)          556   
5   The Hobbit : The Enchanting Prelude to The Lor...          281   
6   The Fellowship of the Ring (The Lord of the Ri...          368   
7   Harry Potter and the Sorcerer's Stone (Harry P...          575   
8      The Two Towers (The Lord of the Rings, Part 2)          260   
9                               To Kill a Mockingbird          510   
10                                  The Da Vinci Code          898   
11                 The Five People You Meet in Heaven          430   
12                             The Catcher in the Rye          449   
13                          The Lovely Bones: A Novel         1295   
14                                               1984          284   
15                           Prodigal Summer: A Novel          253   
16                                         Neverwhere          265   
17                            The Secret Life of Bees          774   
18  Stupid White Men ...and Other Sorry Excuses fo...          283   
19  Tuesdays with Morrie: An Old Man, a Young Man,...          493   
20                The Red Tent (Bestselling Backlist)          723   
21               The Hitchhiker's Guide to the Galaxy          268   
22   Nickel and Dimed: On (Not) Getting By in America          335   
23                                      Into the Wild          252   
24                                     Fahrenheit 451          409   
25                          Girl with a Pearl Earring          526   
26                                          Outlander          283   
27  Where the Heart Is (Oprah's Book Club (Paperba...          585   
28                     Seabiscuit: An American Legend          275   
29                                         Life of Pi          664   
30                                     The Bean Trees          389   
31  A Child Called \It\": One Child's Courage to S...          265   
32                                     ANGELA'S ASHES          279   
33                                        Good in Bed          490   
34  Fast Food Nation: The Dark Side of the All-Ame...          321   
35                                      American Gods          302   
36                                 Skipping Christmas          322   
37  Left Behind: A Novel of the Earth's Last Days ...          318   
38    The Golden Compass (His Dark Materials, Book 1)          336   
39                                   The Color Purple          314   
40                               Silence of the Lambs          256   
41                                        About a Boy          262   
42                  Seven Up (A Stephanie Plum Novel)          278   
43  The Alchemist: A Fable About Following Your Dream          266   
44  Hard Eight : A Stephanie Plum Novel (A Stephan...          269   
45                       Suzanne's Diary for Nicholas          457   
46                                  Lord of the Flies          259   
47                            A Prayer for Owen Meany          413   
48   The Vampire Lestat (Vampire Chronicles, Book II)          301   
49       White Oleander : A Novel (Oprah's Book Club)          356   

    Avg_Rating        ISBN             Book-Author Year-Of-Publication  \
0     5.852804  0439136350           J. K. Rowling                1999   
1     5.824289  0439139597           J. K. Rowling                2000   
2     5.737410  0590353403           J. K. Rowling                1998   
3     5.501441  043935806X           J. K. Rowling                2003   
4     5.183453  0439064872           J. K. Rowling                2000   
5     5.007117  0345339681  

In [102]:
df_books_popular.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Book-Title           50 non-null     object 
 1   Num_Ratings          50 non-null     int64  
 2   Avg_Rating           50 non-null     float64
 3   ISBN                 50 non-null     object 
 4   Book-Author          50 non-null     object 
 5   Year-Of-Publication  50 non-null     object 
 6   Publisher            50 non-null     object 
 7   Image-URL-S          50 non-null     object 
 8   Image-URL-M          50 non-null     object 
 9   Image-URL-L          50 non-null     object 
dtypes: float64(1), int64(1), object(8)
memory usage: 4.0+ KB


In [103]:
final_popular = df_books_popular[['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher', 'Image-URL-L', 'Num_Ratings', 'Avg_Rating']]

* Top 50 Buku Terpopuler

In [104]:
final_popular

ISBN                                         Book-Title  \
0   0439136350  Harry Potter and the Prisoner of Azkaban (Book 3)   
1   0439139597       Harry Potter and the Goblet of Fire (Book 4)   
2   0590353403     Harry Potter and the Sorcerer's Stone (Book 1)   
3   043935806X  Harry Potter and the Order of the Phoenix (Boo...   
4   0439064872   Harry Potter and the Chamber of Secrets (Book 2)   
5   0345339681  The Hobbit : The Enchanting Prelude to The Lor...   
6   0345339703  The Fellowship of the Ring (The Lord of the Ri...   
7   059035342X  Harry Potter and the Sorcerer's Stone (Harry P...   
8   0345339711     The Two Towers (The Lord of the Rings, Part 2)   
9   0446310786                              To Kill a Mockingbird   
10  0385504209                                  The Da Vinci Code   
11  0786868716                 The Five People You Meet in Heaven   
12  0316769487                             The Catcher in the Rye   
13  0316666343                          The Lovely Bones: A Novel   
14  0451524934                                               1984   
15  0060959037                           Prodigal Summer: A Novel   
16  0380789019                                         Neverwhere   
17  0142001740                            The Secret Life of Bees   
18  0060392452  Stupid White Men ...and Other Sorry Excuses fo...   
19  0385484518  Tuesdays with Morrie: An Old Man, a Young Man,...   
20  0312195516                The Red Tent (Bestselling Backlist)   
21  0671461494               The Hitchhiker's Guide to the Galaxy   
22  0805063897   Nickel and Dimed: On (Not) Getting By in America   
23  0385486804                                      Into the Wild   
24  3257208626                                     Fahrenheit 451   
25  0452282152                          Girl with a Pearl Earring   
26  0440222915                                          Outlander   
27  0446672211  Where the Heart Is (Oprah's Book Club (Paperba...   
28  0449005615                     Seabiscuit: An American Legend   
29  0151008116                                         Life of Pi   
30  0060915544                                     The Bean Trees   
31  1558743669  A Child Called \It\": One Child's Courage to S...   
32  0684874350                                     ANGELA'S ASHES   
33  0743418174                                        Good in Bed   
34  0060938455  Fast Food Nation: The Dark Side of the All-Ame...   
35  0380789035                                      American Gods   
36  0385505833                                 Skipping Christmas   
37  0842329129  Left Behind: A Novel of the Earth's Last Days ...   
38  037582345X    The Golden Compass (His Dark Materials, Book 1)   
39  0671617028                                   The Color Purple   
40  0312924585                               Silence of the Lambs   
41  1573227331                                        About a Boy   
42  0312265840                  Seven Up (A Stephanie Plum Novel)   
43  0062502174  The Alchemist: A Fable About Following Your Dream   
44  0312983867  Hard Eight : A Stephanie Plum Novel (A Stephan...   
45  0316969443                       Suzanne's Diary for Nicholas   
46  0399501487                                  Lord of the Flies   
47  0345361792                            A Prayer for Owen Meany   
48  0345313860   The Vampire Lestat (Vampire Chronicles, Book II)   
49  0316284955       White Oleander : A Novel (Oprah's Book Club)   

               Book-Author Year-Of-Publication                      Publisher  \
0            J. K. Rowling                1999                     Scholastic   
1            J. K. Rowling                2000                     Scholastic   
2            J. K. Rowling                1998                     Scholastic   
3            J. K. Rowling                2003                     Scholastic   
4            J. K. Rowling                2000                     Scholastic   
5   

###### Model Development dengan Collaborative Filtering

In [105]:
user_rating = books_rating.groupby('User-ID').count()['Book-Rating']
check_reader = user_rating > 200
experienced_user = user_rating[check_reader].index

- user_rating: Menghitung jumlah rating yang diberikan oleh setiap pengguna (User-ID).
- check_reader: Menentukan pengguna yang telah memberikan lebih dari 200 rating.
- experienced_user: Mendapatkan ID pengguna yang memenuhi syarat tersebut.

In [106]:
rating_filter = books_rating[books_rating['User-ID'].isin(experienced_user)]
rating_filter

User-ID        ISBN  Book-Rating  \
1456      277427  002542730X           10   
1457      277427  0026217457            0   
1458      277427  003008685X            8   
1459      277427  0030615321            0   
1460      277427  0060002050            0   
...          ...         ...          ...   
1147612   275970  3829021860            0   
1147613   275970  4770019572            0   
1147614   275970   896086097            0   
1147615   275970  9626340762            8   
1147616   275970  9626344990            0   

                                                Book-Title  \
1456     Politically Correct Bedtime Stories: Modern Ta...   
1457                    Vegetarian Times Complete Cookbook   
1458                                              Pioneers   
1459      Ask for May, Settle for June (A Doonesbury book)   
1460                     On a Wicked Dawn (Cynster Novels)   
...                                                    ...   
1147612                                     The Penis Book   
1147613                                            Musashi   
1147614                                                NaN   
1147615  Northanger Abbey (Classic Literature with Clas...   
1147616                                                NaN   

                   Book-Author Year-Of-Publication  \
1456         James Finn Garner                1994   
1457                Lucy  Moll                1995   
1458     James Fenimore Cooper                1974   
1459             G. B. Trudeau                1982   
1460         Stephanie Laurens                2002   
...                        ...                 ...   
1147612           Joseph Cohen                1999   
1147613         Eiji Yoshikawa                1995   
1147614                    NaN                 NaN   
1147615            Jane Austen                1996   
1147616                    NaN                 NaN   

                            Publisher  \
1456        John Wiley &amp; Sons Inc   
1457            John Wiley &amp; Sons   
1458                 Thomson Learning   
1459              Henry Holt &amp; Co   
1460                       Avon Books   
...                               ...   
1147612                      Konemann   
1147613  Kodansha International (JPN)   
1147614                           NaN   
1147615         Naxos Audiobooks Ltd.   
1147616                           NaN   

                                               Image-URL-S  \
1456     http://images.amazon.com/images/P/002542730X.0...   
1457     http://images.amazon.com/images/P/0026217457.0...   
1458     http://images.amazon.com/images/P/003008685X.0...   
1459     http://images.amazon.com/images/P/0030615321.0...   
1460     http://images.amazon.com/images/P/0060002050.0...   
...                                                    ...   
1147612  http://images.amazon.com/images/P/3829021860.0...   
1147613  http://images.amazon.com/images/P/4770019572.0...   
1147614                                                NaN   
1147615  http://images.amazon.com/images/P/9626340762.0...   
1147616                                                NaN   

                                               Image-URL-M  \
1456     http://images.amazon.com/images/P/002542730X.0...   
1457     http://images.amazon.com/images/P/0026217457.0...   
1458     http://images.amazon.com/images/P/003008685X.0...   
1459     http://images.amazon.com/images/P/0030615321.0...   
1460     http://images.amazon.com/images/P/0060002050.0...   
...                                                    ...   
1147612  http://images.amazon.com/images/P/3829021860.0...   
1147613  http://images.amazon.com/images/P/4770019572.0...   
1147614                                                NaN   
1147615  http://images.amazon.com/images/P/9626340762.0...   
1147616                                                NaN   

                                               Image-URL-L  
1456     http://images.amazon.

rating_filter: Mengambil rating buku hanya dari pengguna yang dianggap berpengalaman.

In [107]:
book_ratings_count = rating_filter.groupby('Book-Title').count()['Book-Rating']
check_book = book_ratings_count >= 50
popular_books = book_ratings_count[check_book].index

- book_ratings_count: Menghitung jumlah rating untuk setiap buku.
- check_book: Menentukan buku yang telah mendapatkan 50 atau lebih rating.
- popular_books: Mendapatkan judul buku yang memenuhi syarat tersebut.

In [108]:
result_rate = rating_filter[rating_filter['Book-Title'].isin(popular_books)]
result_rate

User-ID        ISBN  Book-Rating  \
1456      277427  002542730X           10   
1469      277427  0060930535            0   
1471      277427  0060934417            0   
1474      277427  0061009059            9   
1480      277427  006440188X            0   
...          ...         ...          ...   
1147440   275970  1400031354            0   
1147441   275970  1400031362            0   
1147517   275970  1573229725            0   
1147556   275970  1586210661            9   
1147557   275970  1586212230            0   

                                                Book-Title  \
1456     Politically Correct Bedtime Stories: Modern Ta...   
1469                         The Poisonwood Bible: A Novel   
1471                                    Bel Canto: A Novel   
1474     One for the Money (Stephanie Plum Novels (Pape...   
1480                                     The Secret Garden   
...                                                    ...   
1147440  Tears of the Giraffe (No.1 Ladies Detective Ag...   
1147441  Morality for Beautiful Girls (No.1 Ladies Dete...   
1147517                                        Fingersmith   
1147556                             Me Talk Pretty One Day   
1147557                                              Naked   

                     Book-Author Year-Of-Publication  \
1456           James Finn Garner                1994   
1469          Barbara Kingsolver                1999   
1471                Ann Patchett                2002   
1474             Janet Evanovich                1995   
1480     Frances Hodgson Burnett                1998   
...                          ...                 ...   
1147440   Alexander McCall Smith                2002   
1147441   Alexander McCall Smith                2002   
1147517             Sarah Waters                2002   
1147556            David Sedaris                2001   
1147557            David Sedaris                2001   

                         Publisher  \
1456     John Wiley &amp; Sons Inc   
1469                     Perennial   
1471                     Perennial   
1474                   HarperTorch   
1480                  HarperTrophy   
...                            ...   
1147440                     Anchor   
1147441                     Anchor   
1147517            Riverhead Books   
1147556    Time Warner Audio Major   
1147557    Time Warner Audio Major   

                                               Image-URL-S  \
1456     http://images.amazon.com/images/P/002542730X.0...   
1469     http://images.amazon.com/images/P/0060930535.0...   
1471     http://images.amazon.com/images/P/0060934417.0...   
1474     http://images.amazon.com/images/P/0061009059.0...   
1480     http://images.amazon.com/images/P/006440188X.0...   
...                                                    ...   
1147440  http://images.amazon.com/images/P/1400031354.0...   
1147441  http://images.amazon.com/images/P/1400031362.0...   
1147517  http://images.amazon.com/images/P/1573229725.0...   
1147556  http://images.amazon.com/images/P/1586210661.0...   
1147557  http://images.amazon.com/images/P/1586212230.0...   

                                               Image-URL-M  \
1456     http://images.amazon.com/images/P/002542730X.0...   
1469     http://images.amazon.com/images/P/0060930535.0...   
1471     http://images.amazon.com/images/P/0060934417.0...   
1474     http://images.amazon.com/images/P/0061009059.0...   
1480     http://images.amazon.com/images/P/006440188X.0...   
...                                                    ...   
1147440  http://images.amazon.com/images/P/1400031354.0...   
1147441  http://images.amazon.com/images/P/1400031362.0...   
1147517  http://images.amazon.com/images/P/1573229725.0...   
1147556  http://images.amazon.com/images/P/1586210661.0...   
1147557  http://images.amazon.com/images/P/1586212230.0...   

                                               Image-URL-L  
1456     http://images.amazon.com/images/P

result_rate: Mengambil rating hanya untuk buku yang termasuk dalam kategori buku populer.

In [109]:
pivot_book_ratings = result_rate.pivot_table(index='Book-Title', columns='User-ID', values='Book-Rating')
pivot_book_ratings

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    3757    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4017    4385    6242    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     7.0   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6251    ...  274004  \
Book-Title                                                  ...           
1984                                                   NaN  ...     NaN   
1st to Die: A Novel                                    NaN  ...     NaN   
2nd Chance                                             NaN  ...     NaN   
4 Blondes                                              0.0  ...     NaN   
84 Charing Cross Road                                  NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        NaN  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

In [110]:
pivot_book_ratings.fillna(0, inplace=True)
pivot_book_ratings

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    3757    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4017    4385    6242    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     7.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6251    ...  274004  \
Book-Title                                                  ...           
1984                                                   0.0  ...     0.0   
1st to Die: A Novel                                    0.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
84 Charing Cross Road                                  0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

- pivot_book_ratings: Membuat tabel pivot dengan judul buku sebagai indeks, ID pengguna sebagai kolom, dan rating buku sebagai nilai.
- fillna(0): Mengisi nilai kosong dengan 0.

In [111]:
from sklearn.metrics.pairwise import cosine_similarity
similarity_score = cosine_similarity(pivot_book_ratings)
similarity_score.shape

(742, 742)

similarity_score: Menghitung kemiripan antara buku menggunakan cosine similarity. Hasilnya adalah matriks kemiripan dengan ukuran yang sama dengan jumlah buku.

In [112]:
import numpy as np

In [113]:
def recommendation(book_name):
    # Temukan index buku yang diminta dalam pivot_book_ratings
    index = np.where(pivot_book_ratings.index == book_name)[0][0]

    # Menghitung skor kemiripan dan memilih 5 item teratas, kecuali item itu sendiri
    similar_item = sorted(list(enumerate(similarity_score[index])), key=lambda x: x[1], reverse=True)[1:6]

    # Menyimpan hasil rekomendasi dalam list
    recommendations = []
    for i in similar_item:
        # Ambil data buku yang sesuai dari books_fix
        temp_df = books_fix[books_fix['Book-Title'] == pivot_book_ratings.index[i[0]]]
        for _, row in temp_df.iterrows():
            recommendations.append({
                'Book-Title': row['Book-Title'],
                'Book-Author': row['Book-Author'],
                'Book-Publisher': row['Publisher']
            })

    # Membuat DataFrame dari hasil rekomendasi
    recommendations_df = pd.DataFrame(recommendations)

    return recommendations_df

- recommendation(book_name): Fungsi yang memberikan rekomendasi buku serupa berdasarkan nama buku yang diberikan.
  - index: Menentukan indeks buku yang dimaksud dalam tabel pivot.
  - similar_item: Mendapatkan buku-buku yang paling mirip dengan buku yang dimaksud berdasarkan kemiripan cosine.
  - data: Mengumpulkan data judul dan penulis buku yang direkomendasikan.

In [114]:
result_df = recommendation('Year of Wonders')
result_df

Book-Title        Book-Author  \
0                                  The Eight  Katherine Neville   
1  A Heartbreaking Work of Staggering Genius        Dave Eggers   
2                               Empire Falls      Richard Russo   
3          Bridget Jones: The Edge of Reason     Helen Fielding   
4                             A Civil Action      JONATHAN HARR   

      Book-Publisher  
0   Ballantine Books  
1  Vintage Books USA  
2  Vintage Books USA  
3      Penguin Books  
4            Vintage

Output dari kode diatas akan menampilkan 5 rekomendasi buku sesuai syarat yang ditentukan

###### Evaluation

Metrik evaluasi yang digunakan adalah Mean Average Precision (MAP). MAP (Mean Average Precision) mengukur kinerja sistem rekomendasi dengan menghitung rata-rata dari precision di berbagai posisi relevan dalam daftar rekomendasi. Precision diukur pada setiap posisi relevan dalam daftar rekomendasi dan kemudian dirata-ratakan untuk seluruh item.

In [129]:
def evaluate_map(recommended_books, threshold, k):
    # Daftar untuk menyimpan precision pada posisi relevan
    avg_precision_list = []

    # Grup data berdasarkan item untuk evaluasi
    grouped = recommended_books.groupby('Book-Title')

    for title, group in grouped:
        relevant_items = group['Book-Rating'] > threshold
        retrieved = np.sum(relevant_items)

        if retrieved == 0:
            continue

        precision_at_i = [np.sum(relevant_items[:i+1]) / (i+1) for i in range(len(relevant_items))]
        avg_precision = np.sum(precision_at_i * relevant_items) / np.sum(relevant_items)
        avg_precision_list.append(avg_precision)

    map_score = np.mean(avg_precision_list) if avg_precision_list else 0
    return map_score

# Contoh penggunaan
k = 10
threshold = 2
map_score = evaluate_map(rating_filter, threshold, k)
print(f'The MAP score of the recommendation system is {map_score:.2%}')


The MAP score of the recommendation system is 78.51%


- evaluate_map(recommended_books, threshold, k): Fungsi ini menghitung Mean Average Precision (MAP) dari buku yang direkomendasikan. Parameter:
recommended_books: Dataframe yang berisi buku-buku yang direkomendasikan dengan ratingnya.
threshold: Nilai rating yang dianggap relevan.
k: Jumlah rekomendasi yang diambil (tidak digunakan secara eksplisit dalam kode, tetapi dapat mempengaruhi pengaturan data)

- avg_precision_list = []: Daftar untuk menyimpan nilai precision rata-rata untuk setiap buku.

- grouped = recommended_books.groupby('Book-Title'): Mengelompokkan data berdasarkan judul buku (Book-Title), sehingga setiap kelompok berisi rating yang diberikan untuk buku tertentu.

- for title, group in grouped:: Iterasi melalui setiap buku dan kelompoknya.
relevant_items = group['Book-Rating'] > threshold: Menentukan item yang relevan berdasarkan threshold. Ini menghasilkan array boolean yang menunjukkan apakah rating suatu buku lebih besar dari threshold.
retrieved = np.sum(relevant_items): Menghitung jumlah item relevan.
if retrieved == 0: continue: Jika tidak ada item relevan, lanjutkan ke buku berikutnya.
precision_at_i = [np.sum(relevant_items[:i+1]) / (i+1) for i in range(len(relevant_items))]: Menghitung precision pada setiap posisi i. Precision pada posisi i adalah proporsi item relevan yang ditemukan sampai posisi i.
avg_precision = np.sum(precision_at_i * relevant_items) / np.sum(relevant_items): Menghitung rata-rata precision untuk buku tersebut, dengan mempertimbangkan hanya posisi relevan.

- map_score = np.mean(avg_precision_list) if avg_precision_list else 0: Menghitung MAP sebagai rata-rata dari precision rata-rata untuk semua buku. Jika tidak ada buku, MAP diatur ke 0.

- map_score = evaluate_map(rating_filter, threshold, k): Menghitung MAP berdasarkan data rekomendasi (rating_filter), threshold, dan jumlah rekomendasi k.
print(f'The MAP score of the recommendation system is {map_score:.2%}'): Menampilkan hasil MAP sebagai persentase.